pennylane/workflow/construct_batch.py

```diff
@@ -355,14 +457,114 @@ def batch_constructor(*args, **kwargs) -> tuple[QuantumScriptBatch, Postprocessi
+            # If the level slice is fully contained within user transforms, we can return early
+            return program((initial_tape,))
+
+        user_transformed_tapes, user_post_processing = program((initial_tape,))
+        #### User transforms finished #####
+        # The new config process we would like to use.
+        mcm_config = qml.devices.MCMConfig(
+            postselect_mode=qnode.execute_kwargs["postselect_mode"],
+            mcm_method=qnode.execute_kwargs["mcm_method"],
         )
-        gradient_fn = config.gradient_method
-        program = get_transform_program(qnode, level=level, gradient_fn=gradient_fn)
+        execution_config = _make_execution_config(
+            qnode, qnode.diff_method, mcm_config
+        )  # pylint: disable = protected-access
+
+        ###### Resolution of the execution config ######
+        execution_config = _resolve_execution_config(
+            execution_config,
+            qnode.device,
+            tapes=user_transformed_tapes,  # Use the user-transformed tapes
+        )
+
+        # Use _setup_transform_program like execute() does
+        outer_transform_program, inner_transform_program = _setup_transform_program(
+            qnode.device,
+            execution_config,
+            cache=qnode.execute_kwargs["cache"],
+            cachesize=qnode.execute_kwargs["cachesize"],
+        )
+        full_transform_program = outer_transform_program + inner_transform_program
+
+        gradient_fn = execution_config.gradient_method
+        has_gradient_expand = bool(
+            getattr(gradient_fn, "expand_transform", False)
+        )  # Note that it could exist as None which is still False, but can't use hasattr on it.
+        level_slice_inner = _interpret_level_inner(
+            level,
+            num_user_transforms,
+            has_gradient_expand,
+            qnode.transform_program.has_final_transform,
+        )
+        shift_len = num_user_transforms - int(qnode.transform_program.has_final_transform)
+
+        def _shift(i):
+            return i - shift_len if i else i
+
+        new_start = _shift(level_slice_inner.start)
+        new_stop = _shift(level_slice_inner.stop)
 
-        return program((initial_tape,))
+        level_slice_inner = slice(new_start, new_stop, level_slice_inner.step)
+        resolved_program = full_transform_program[level_slice_inner]
+
+        batch, remaining_post_processing = resolved_program(
+            user_transformed_tapes
+        )  # Use the user-transformed tapes
+
+        def combined_post_processing(results):
+            """Combine the user post-processing with the remaining post-processing."""
+            intermediate_results = remaining_post_processing(results)
+            return user_post_processing(intermediate_results)
+
+        return batch, combined_post_processing
 
     return batch_constructor
+
+
+def _is_within_user_bounds(level, num_user: int) -> bool:
```

Comment:
Don't need this function anymore. We should be thinking in generalities, not specific cases.  Left suggestion above where we just replace it with `if initial_level_slice.stop is not None and initial_level_slice.stop <= num_user)`